### PR TITLE
Handle closure return types

### LINKED ISSUE #2547: Type of closures with the return statement is inferred incorrectly
A spin-off issue of #2534 

```rust
fn main() {
    let foo = || -> usize {
        return 1;
    };
}
```

The inferred type of `foo` is `|| -> ()` instead of `|| -> usize`.

### PR DESCRIPTION
**Pull request by flodiebold**
Fixes #2547.

### CODE REVIEW

File: crates/ra_hir_ty/src/infer.rs [hunk 1]
```diff
@@ -451,11 +458,15 @@ impl<'a, D: HirDatabase> InferenceContext<'a, D> {
...
-        self.return_ty = self.insert_vars_for_impl_trait(return_ty);
+        let return_ty = self.insert_vars_for_impl_trait(return_ty);
+        self.return_tys.push(return_ty);
     }
 
     fn infer_body(&mut self) {
-        self.infer_expr(self.body.body_expr, &Expectation::has_type(self.return_ty.clone()));
+        self.infer_expr(
+            self.body.body_expr,
+            &Expectation::has_type(self.return_tys.last().cloned().unwrap_or(Ty::Unknown)),
```
Comment (matklad) [inline crates/ra_hir_ty/src/infer.rs]:
```suggestion
            &Expectation::has_type(self.return_tys.last().unwrap().clone()),
```

I think we should guarantee that the stack has at least one element?

File: crates/ra_hir_ty/src/infer/expr.rs [hunk 2]
```diff
@@ -134,7 +137,12 @@ impl<'a, D: HirDatabase> InferenceContext<'a, D> {
                 // infer the body.
                 self.coerce(&closure_ty, &expected.ty);
 
-                self.infer_expr(*body, &Expectation::has_type(ret_ty));
+                self.return_tys.push(ret_ty.clone());
+
+                self.infer_expr_coerce(*body, &Expectation::has_type(ret_ty));
+
+                self.return_tys.pop();
```
Comment (matklad) [inline crates/ra_hir_ty/src/infer/expr.rs]: Hm, do we actually need a stack of return types? We already have a call stack!

```suggestion
let prev_ret_ty = mem::replace(&mut self.ret_ty, ret_ty.clone())
self.infer_expr_coerce(*body, &Expectation::has_type(ret_ty));
self.ret_ty = prev_ret_ty;
```

Comment (flodiebold) [reply crates/ra_hir_ty/src/infer/expr.rs]: oh, yeah, of course

**Pull request merged**